File: src/main/java/com/thomaskint/minidao/annotation/MDManyToOne.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.thomaskint.minidao.annotation;

import com.thomaskint.minidao.enumeration.MDLoadPolicy;
import com.thomaskint.minidao.utils.MDIncomingFeature;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import static com.thomaskint.minidao.enumeration.MDLoadPolicy.LAZY;

/**
 * @author Thomas Kint
 */
@Target(ElementType.FIELD)
@Retention(RetentionPolicy.RUNTIME)
public @interface MDManyToOne {

	String fieldName();

	Class target();

	@MDIncomingFeature(
			featureName = "Control of load policy",
			description = "The load of the parent target will be affected by that value")
	MDLoadPolicy loadPolicy() default LAZY;
}

File: src/main/kotlin/com/thomaskint/minidao/annotation/MDManyToOne.kt
/*
 * MIT License
 *
 * Copyright (c) 2017 Thomas Kint
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.thomaskint.minidao.annotation

import com.thomaskint.minidao.enumeration.MDLoadPolicy
import com.thomaskint.minidao.enumeration.MDLoadPolicy.LAZY
import com.thomaskint.minidao.utils.MDIncomingFeature
import kotlin.reflect.KClass

/**
 * @author Thomas Kint
 */
@Target(AnnotationTarget.FIELD)
@kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
annotation class MDManyToOne(
        val fieldName: String,

        val targetFieldName: String,

        val target: KClass<*>,

        @MDIncomingFeature(featureName = "Control of load policy", description = "The load of the target will be affected by that value")
        val loadPolicy: MDLoadPolicy = LAZY
)
